### PR TITLE
Persist uuid and name in device memory.

### PR DESCRIPTION
**Pull request by cor3ntin**
Not really tested yet, pushing for early review

### CODE REVIEW

File: skel-usb.c [hunk 1]
```diff
@@ -295,37 +279,60 @@ void AsebaWriteBytecode(AsebaVMState *vm) {
...
+    page_addr += INSTRUCTIONS_PER_ROW * 2 * i;
 	
-	flash_prepare_write(setting_addr);
-	temp = (((unsigned long) *((unsigned char * ) &settings)) << 16) | _magic_[i];
+    flash_prepare_write(page_addr);
+    unsigned long temp = (((unsigned long) *((unsigned char * ) data)) << 16) | _magic_[i];
 	
 	flash_write_instruction(temp);
-	flash_write_buffer(((unsigned char *) &settings) + 1, sizeof(settings) - 1);
+    flash_write_buffer(((unsigned char *) data) + 1, size - 1);
```
Comment (mbonani) [inline skel-usb.c]: be consistent with indentation

**Comment by mbonani** [comment]
at first lecture seems to be testable

File: skel-usb.c [hunk 2]
```diff
@@ -508,4 +513,69 @@ void save_settings(void) {
...
\ No newline at end of file
+}
+//Called from the vm, escape hatch to handle messages specific to thymio.
+//Returns true if a message was handled
+int AsebaHandleThymioSpecificMessage(AsebaVMState* vm, uint16_t id, uint16_t* data, uint16_t dataLength) {
+    if(id == ASEBA_MESSAGE_THYMIO_GET_THYMIO_DEVICE_INFO || id == ASEBA_MESSAGE_THYMIO_SET_THYMIO_DEVICE_INFO) {
+        uint8_t* bytes = (uint8_t*)data;
+        uint16_t size  = dataLength * 2;
+        if(size < 1) // We need at least an info type
+            return;
```
Comment (retornaz) [inline skel-usb.c]: Enable or check compiler warning.

File: skel-usb.c [hunk 3]
```diff
@@ -508,4 +513,69 @@ void save_settings(void) {
...
+                AsebaSendMessage(vm, ASEBA_MESSAGE_THYMIO_DEVICE_INFO, payload, payload_size);
+            }
+        }
+        else if (id == ASEBA_MESSAGE_THYMIO_SET_THYMIO_DEVICE_INFO) {
+            if(size < 2) // We need at least an info type (1) + size (2)
+                return FALSE;
+            uint8_t payload_size  = bytes[2];
+            const uint8_t* buffer = bytes + 3;
+            if(size > (ASEBA_MAX_PACKET_SIZE + 6) || payload_size < size + 2) {
+                return FALSE;
```
Comment (retornaz) [inline skel-usb.c]: 0/1 is more in the spirit of the code IIRC

File: skel-usb.h [hunk 4]
```diff
@@ -85,6 +85,11 @@ void update_aseba_variables_write(void);
 // return non-zero if the settings were NOT found (settings is non-initilised)
 int load_settings_from_flash(void);
 
+// Loads some persistent informations specific to thymio 2
+// This are never used by the device itself but are used by applications using the device.
+// It includes a unique uuid ( initialized on a computer ) as well as a device name that user can modify
+int load_thymio_device_info_from_flash();
```
Comment (retornaz) [inline skel-usb.h]: void

File: skel-usb.c [hunk 5]
```diff
@@ -508,4 +513,69 @@ void save_settings(void) {
...
+        uint8_t* bytes = (uint8_t*)data;
+        uint16_t size  = dataLength * 2;
+        if(size < 1) // We need at least an info type
+            return;
+        uint8_t type = bytes[0];
+        if(type > THYMIO_DEVICE_INFO_ENUM_COUNT) // Send an error ?
+            return;
+
+        if (id == ASEBA_MESSAGE_THYMIO_GET_THYMIO_DEVICE_INFO) {
+            uint8_t size = 0;
```
Comment (retornaz) [inline skel-usb.c]: so you have two size variable with two different type and meaning ... no way.

Comment (cor3ntin) [reply skel-usb.c]: Size of the message, and size of the variable we are setting. Although "payload_size" is a poor choice of name, I'll change that.

**Comment by cor3ntin** [comment]
(Fair warning: This code is *not* working, a thymio died while testing)

**Comment by mbonani** [comment]
is this now working?

**Comment by cor3ntin** [comment]
Yep

**Pull request merged**